### ISSUE TITLE
Cohorts take too long to load

### ISSUE THREAD

**Issue by Tannergoods** [PostHog/posthog#333]
***Thank*** you for reporting bugs - we appreciate every single bug report.

**Describe the bug**
My cohorts won't load up.

**To Reproduce**
1. Go to 'app.posthog.com/people/cohorts'
2. Click on one cohort or create a new one that will likely have more than 40+ users
3. See error - it won't load

**Expected behavior**
Cohorts should load in a reasonable timeframe

**Screenshots**
<img width="1145" alt="Posthog" src="https://user-images.githubusercontent.com/60791437/76473828-eb3c0b80-63b6-11ea-9295-09be01e4a47d.png">

**Desktop (please complete the following information):**
 - MacOS
 - chrome

**Comment by timgl** [comment]
The reason for this is that when the action filters on 'Contains URL', we do a LIKE query (`v0.properties ->> '$current_url' LIKE '%.herokuapp.com%'`) over every single row in the events table. I think this query literally never finishes.

I'm not sure what the best approach to solving this is. You can't index LIKE queries, but I do think this is an important feature to have. @mariusandra Do you have any ideas on how to solve this?

**Comment by timgl** [comment]
@Tannergoods I did put a PR in #339 that fixes a bug that meant there was an error when you selected "in the last ...".

**Comment by mariusandra** [comment]
Hi @timgl and others!

I guess this is the query in question? (With a different URL of course)

```sql
SELECT "posthog_person"."id",
       "posthog_person"."created_at",
       "posthog_person"."team_id",
       "posthog_person"."properties",
       "posthog_person"."is_user_id",
       (
        SELECT V0."id"
          FROM "posthog_event" V0
         WHERE (V0."team_id" = 1 AND V0."event" = '$pageview' AND UPPER((V0."properties" ->> '$current_url')::text) LIKE UPPER('%/demo/2%') AND (SELECT U0."person_id" FROM "posthog_persondistinctid" U0 WHERE (U0."distinct_id" = V0."distinct_id" AND U0."team_id" = 1) LIMIT 1) = "posthog_person"."id")
         ORDER BY V0."id" DESC
         LIMIT 1
       ) AS "has_action"
  FROM "posthog_person"
 WHERE ("posthog_person"."team_id" = 1 AND (SELECT V0."id" FROM "posthog_event" V0 WHERE (V0."team_id" = 1 AND V0."event" = '$pageview' AND UPPER((V0."properties" ->> '$current_url')::text) LIKE UPPER('%/demo/2%') AND (SELECT U0."person_id" FROM "posthog_persondistinctid" U0 WHERE (U0."distinct_id" = V0."distinct_id" AND U0."team_id" = 1) LIMIT 1) = "posthog_person"."id") ORDER BY V0."id" DESC LIMIT 1) IS NOT NULL)
```

This is an interesting situation. I honestly don't know the best path forward, but here are some ideas.

First, you do have an index that will help with LIKE queries. This answer explains more: https://stackoverflow.com/a/13452528

It'll probably be a huge index, but running this *should* give you some performance benefits:

```sql
CREATE EXTENSION pg_trgm;
CREATE INDEX CONCURRENTLY posthog_event_properties_current_url_gin ON posthog_event USING gin (UPPER(properties->>'$current_url') gin_trgm_ops);
```

I made a migration for it in PR #377

Unfortunately I had to create a custom SQL migration as [django doesn't support creating functional indexes](https://github.com/django/django/pull/11929)... so there's nothing about this index in `models.py`.

The index seems to do something though. Running `EXPLAIN ANALYZE SELECT` shows the following change:

Before:
![image](https://user-images.githubusercontent.com/53387/76976081-bd654280-6933-11ea-8c28-8e87a73cff83.png)

After:
![image](https://user-images.githubusercontent.com/53387/76976137-d968e400-6933-11ea-9ef9-f3c053616816.png)

However, even if this index helps a bit now, as your or your clients' databases grow, it might not be enough and things will get slow again eventually.

In that case, if you intend to continue using postgresql, there's only one real solution: pre-calculating data. In this case it would mean creating a has-many table (e.g. posthog_event_actions) or an array[] field on posthog_events to store all the actions that each event matches. Then whenever a new event is added, run a background job to update all the events and add the corresponding action to it.

As you can imagine, this approach will have problems of its own... but may be necessary eventually :).

**Comment by timgl** [comment]
Will be solved in #696
